grep-0021-pmt-replacement.md

```diff
@@ -0,0 +1,90 @@
+# GREP 0019 -- Restructure GNU Radio Blocks
+
+- Original Author: Josh Morman <mormjb@gmail.com>
+- Champion: Josh Morman <mormjb@gmail.com>
+- Status: Draft
+
+History:
+
+- 25-Feb-2021: Initial Draft
+
+## Abstract
+
+This GREP proposes replacing the PMT types with classes that support better 
+serialization and have a cleaner API
+
+## Copyright / License
+
+CC-BY-ND
+
+## Motivation
+
+The existing PMT API is confusing, doesn't take advantage of OOP concepts, and
+the nomenclature for doing simple things can be inconsistent or lacking.  Additionally,
+there has been work to show that the serialization/deserialization is inherently slow
+and specific to the PMT library.
+
+The leading candidate to natively support serialization of polymorphic types is
+[flatbuffers](https://google.github.io/flatbuffers/).  This library provides cross
+platform serialization and deserialization and is efficient in storing the serialized
+representation in a flat buffer that can be accessed directly
+
+## Description
+
+Create a new in-tree module - `pmtf` - that will live alongside `pmt` during a 
+deprecation cycle.  The `pmtf` namespace will have a drastically updated API that
+is more similar to `std::` and other libraries used inside the GNU Radio codebase.
+
+`pmtf` will wrap flatbuffer representations of polymorphic objects (underlying
+blob can map to a fixed number of enumerated types)
+
+The work to prototype flatbuffer based pmts is being prototyped [here](https://github.com/gnuradio/newsched/tree/master/pmt)
+
+### Proposed API (looking for feedback)
+
+Templated classes derived from `pmt_base` represent the _container_ holding the specified
+type:
+
+- `pmt_scalar`
+  - represents a single value (e.g. `pmt_scalar<uint8_t>`)
+- `pmt_vector`
+  - represents a uniform vector (e.g. `pmt_vector<gr_complex>`)
+- `pmt_map`
+  - roughly replaces `pmt::dict`, but the generality of the anything-to-anything
+  dict is hard to serialize with flatbuffers.  `map` seems to be a reasonable simplification
+- `pmt_string`
+  - a string
+
+
+Objects (same as in legacy pmts) are generated as a shared pointer to a PMT object, 
+since they are meant to be passed around
+
+```
+auto pmt_val = pmt_scalar<std::complex<float>>::make(cplx_val);
+
+std::vector<int32_t> int_vec_val{ 5, 9, 23445, 63, -25 };
+auto int_pmt_vec = pmt_vector<int32_t>::make(int_vec_val);
+```
+
+Maps are created by specifying the key type, but the value is always another PMT:
+```
+// Create the PMT map
+std::map<std::string, pmt_sptr> input_map({
+  { "key1", pmt_scalar<std::complex<float>>::make(val1) },
+        { "key2", pmt_vector<int32_t>::make(val2) },
+  });
+auto map_pmt = pmt_map<std::string>::make(input_map);
+```
+
+### Additional Dependencies
+
+The flatbuffers library and flatc compiler will need to be included to compile the PMT schema definition into a generated header
+
+Alternatively if we don't expect the schema to change very ofted, this can be done as a
```

Comment:
Is this something we could manage similar to pybind bindings? I would not expect this to change much once it is integrated, libpmt has changed relatively little its entire life so far, but CMake would need a way to clearly alert users.

Reply:
Yeah I think that would be a safe approach (method of manual generation in-tree).  What I've noticed so far with flatbuffers is that the generation using flatc can vary from version to version especially with complex data types.  So we would want to ensure that the in-tree generation with an optional dependency of a controlled version of flatc.

Reply:
Yes, experience from protocol buffers probably applies to flatbuffers: never save the generated code.

Reply:
So would your recommendation be assuming that flatbuffers over a range of compatible versions will generate the necessary code?  How would you suggest managing the dependency?

Reply:
The flatbuffers API seems very stable, from a quick look through their changelogs. I think the risk of using different versions of flatc is low, there won't be miles of incomprehensible code checked in, and we get any improvements with new version.

So, how about any version >= whatever is currently in various distributions (11.1+?) is allowed now. If there's an incompatible change in their API in the future, revisit and possibly bump the requirement.

We can also encourage simplest possible use of the flatbuffers API. If GR's use of flatbuffers will be wrapped in GR functionality (e.g., PMT), then the usage won't get out of hand. If we open up flatbuffers as an external message interface, then a little more guidance could be needed to keep people from using every available feature.

Reply:
Generation from flatc does vary quite a lot, but the FlatBuffer folks have strong promises around cross-compatibility. If we wrap the flatbuffers as opposed to exposing them directly, then I see little risk here allowing a range of flatbuffer-versions.